// jest-cli/.haste_cache/preprocess-cache/time-services_5668209deb2409032ba0c11a0198f5f2.js
'use strict';

var Moment = require('moment');
var Validate = require('validate-arguments');


var TimeServices = function () {

  var months = Moment.months();

  var now = new Moment();

  return { 
    convertFromMinutesAfterMidnight: function /*istanbul ignore next*/convertFromMinutesAfterMidnight(offsetInMinutes) {
      if (typeof offsetInMinutes === 'undefined') {
        throw new Error('function in time-services called with no parameters');}

      var hour = Math.floor(offsetInMinutes / 60);
      var minute = offsetInMinutes % 60;

      return { 
        hour: hour, 
        minute: minute };}, 



    convertFromDaysAfterNewYear: function /*istanbul ignore next*/convertFromDaysAfterNewYear(offsetInDays, year) {
      if (typeof offsetInDays === 'undefined') {
        throw new Error('function in time-services called with no parameters');}

      checkForEmptyArgs(offsetInMinutes);
      if (typeof year === 'undefined') {
        year = now.year();}

      var newYear = Moment(new Date(year, 0, 1));
      var newDate = newYear.add(offsetInDays, 'days');
      return { 
        year: newDate.year(), 
        month: newDate.month(), 
        date: newDate.date() };}, 



    getMinutesSinceMidnight: function /*istanbul ignore next*/getMinutesSinceMidnight(time) {
      validateTime(time);

      return time.hour * 60 + time.minute;}, 


    getDaysSinceNewYear: function /*istanbul ignore next*/getDaysSinceNewYear(date) {

      validateDate(date);

      var newYear = Moment(new Date(date.year, 0, 1));
      var current = Moment(new Date(date.year, date.month, date.date));
      return current.diff(newYear, 'days');}, 


    formattedTime: function /*istanbul ignore next*/formattedTime(time) {

      validateTime(time);

      var timeString = '', 
      ampm = 'am', 
      hour = time.hour, 
      minute = time.minute;

      if (hour >= 12) ampm = 'pm';
      if (hour > 12) hour = hour - 12;
      if (hour === 0) {
        hour = 12;
        ampm = 'am';}


      hour = hour.toString();
      minute = minute.toString();

      if (minute < 10) minute = '0' + minute;

      timeString += hour + ':' + minute + ' ' + ampm;

      return timeString;}, 


    formattedDate: function /*istanbul ignore next*/formattedDate(date) {

      validateDate(date);

      var dateString = '';
      dateString += months[date.month] + ' ' + date.date + ', ' + date.year;

      return dateString;} };}();







function validateTime(time) {

  checkForEmptyArgs(time);

  var args = Validate.validate(time, { 

    hour: { 
      isa: 'whole' }, 

    minute: { 
      isa: 'whole' } });



  if (!args.isValid()) {
    throw args.errorString();}}



function validateDate(date) {

  checkForEmptyArgs(date);

  var args = Validate.validate(date, { 

    year: { 
      isa: 'natural' }, 

    month: { 
      isa: 'natural' }, 

    date: { 
      isa: 'natural' } });



  if (!args.isValid()) {
    throw args.errorString();}}



function checkForEmptyArgs(args) {
  if (typeof args === 'undefined') {
    throw new Error('function in time-services called with no parameters');}}



module.exports = TimeServices;
