// jest-cli/.haste_cache/preprocess-cache/validate_6a22868ff32a47fa454de0805bdf4272.js
/**
 *                                        validate.js
 *
 *    This module is meant to validate function parameters as a service for APIs in the application.  It's not meant to
 *    verify user input.  Here, errors will be thrown if something doesn't validate.  The module is designed to cause
 *    my programming mistakes to crash the application.
 *
 *    For simple stuff, some basic functions are provided.  For JavaScript object validation, Joi is used.
 *
 */

'use strict';

var Joi = require('joi');

var Validate = function () {

  return { 

    hasArguments: function /*istanbul ignore next*/hasArguments(args) {
      return _hasArguments(args);}, 


    isNumber: function /*istanbul ignore next*/isNumber(arg) {
      return _isNumber(arg);}, 


    isWithinRange: function /*istanbul ignore next*/isWithinRange(value, min, max) {
      return _isWithinRange(value, min, max);}, 


    time: function /*istanbul ignore next*/time(_time) {
      return validateTime(_time);}, 


    date: function /*istanbul ignore next*/date(_date) {
      return validateDate(_date);}, 


    latitude: function /*istanbul ignore next*/latitude(_latitude) {
      return validateLatitude(_latitude);}, 


    longitude: function /*istanbul ignore next*/longitude(_longitude) {
      return validateLongitude(_longitude);} };



  function validateTime(time) {

    _hasArguments(time.hour);
    _hasArguments(time.minute);

    var schema = Joi.object().keys({ 
      hour: Joi.number().integer().min(0).max(23).required(), 
      minute: Joi.number().integer().min(0).max(59).required() });


    Joi.validate(time, schema, function (err) {
      if (err) throw err;});}



  function validateDate(date) {

    _hasArguments(date);

    var schema = Joi.object().keys({ 
      year: Joi.number().integer().min(-9999).max(9999).required(), 
      month: Joi.number().integer().min(0).max(11).required(), 
      date: Joi.number().integer().min(1).max(31).required() });


    Joi.validate(date, schema, function (err) {
      if (err) throw err;});}




  function validateLatitude(latitude) {

    _hasArguments(latitude);

    if (!_isWithinRange(latitude, -90, 90)) {
      throw new Error('bad arguments');}}



  function validateLongitude(longitude) {

    _hasArguments(longitude);

    _isWithinRange(longitude, -180, 180);}


  function _hasArguments(args) {
    if (typeof args !== 'undefined') {
      return true;} else 
    {
      throw new Error('missing arguments');}}



  function _isNumber(arg) {
    if (!isNaN(arg)) {
      return true;} else 
    {
      throw new Error('argument is not a number');}}



  function _isWithinRange(value, min, max) {
    if (value >= min && value <= max) {
      return true;} else 
    {
      throw new Error('argument out of range');}}}();





module.exports = Validate;
