// jest-cli/.haste_cache/preprocess-cache/viewport-actions_924d139a48c229f84ae2c71d6504b81e.js
/*istanbul ignore next*/'use strict';var AppDispatcher = require('../dispatcher/app-dispatcher');
var ViewportConstants = require('../constants/viewport-constants');

var ViewportActions = { 

  createObject: function /*istanbul ignore next*/createObject(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.CREATE_OBJECT, 
      objectData: data.objectData });}, 



  setTime: function /*istanbul ignore next*/setTime(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.SET_TIME, 
      time: data.time });}, 



  setDate: function /*istanbul ignore next*/setDate(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.SET_DATE, 
      date: data.date });}, 



  setLatitude: function /*istanbul ignore next*/setLatitude(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.SET_LATITUDE, 
      latitude: data.latitude });}, 



  setLongitude: function /*istanbul ignore next*/setLongitude(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.SET_LONGITUDE, 
      longitude: data.longitude });}, 



  saveCameraData: function /*istanbul ignore next*/saveCameraData(data) {
    AppDispatcher.dispatch({ 
      actionType: ViewportConstants.SAVE_CAMERA_DATA, 
      cameraData: data.cameraData });} };




module.exports = ViewportActions;
